modules/http4s-server/src/main/scala/io/janstenpickle/trace4cats/http4s/server/Http4sResourceKleislis.scala

```diff
@@ -1,35 +1,49 @@
 package io.janstenpickle.trace4cats.http4s.server
 
-import cats.{Applicative, Monad}
 import cats.data.Kleisli
 import cats.effect.Resource
-import io.janstenpickle.trace4cats.Span
-import io.janstenpickle.trace4cats.http4s.common.{Http4sHeaders, Http4sRequestFilter, Http4sSpanNamer, Request_}
+import cats.{Applicative, Monad}
+import io.janstenpickle.trace4cats.http4s.common.{
+  Http4sHeaders,
+  Http4sRequestFilter,
+  Http4sSpanNamer,
+  Http4sStatusMapping,
+  Request_
+}
 import io.janstenpickle.trace4cats.inject.{ResourceKleisli, SpanParams}
 import io.janstenpickle.trace4cats.model.SpanKind
-import org.http4s.Headers
+import io.janstenpickle.trace4cats.{ErrorHandler, HandledError, Span}
+import org.http4s.{Headers, HttpVersion, MessageFailure}
 import org.http4s.util.CaseInsensitiveString
 
 object Http4sResourceKleislis {
+  private val messageFailureHandler: ErrorHandler = { case e: MessageFailure =>
+    HandledError.Status(Http4sStatusMapping.toSpanStatus(e.toHttpResponse(HttpVersion.`HTTP/1.0`).status))
+  }
+
   def fromHeadersContext[F[_]: Monad, Ctx](
     makeContext: (Request_, Span[F]) => F[Ctx],
     spanNamer: Http4sSpanNamer = Http4sSpanNamer.methodWithPath,
     requestFilter: Http4sRequestFilter = Http4sRequestFilter.allowAll,
     dropHeadersWhen: CaseInsensitiveString => Boolean = Headers.SensitiveHeaders.contains,
+    errorHandler: ErrorHandler = ErrorHandler.empty,
   )(k: ResourceKleisli[F, SpanParams, Span[F]]): ResourceKleisli[F, Request_, Ctx] =
-    fromHeaders[F](spanNamer, requestFilter, dropHeadersWhen)(k).tapWithF { (req, span) =>
+    fromHeaders[F](spanNamer, requestFilter, dropHeadersWhen, errorHandler)(k).tapWithF { (req, span) =>
       Resource.liftF(makeContext(req, span))
     }
 
   def fromHeaders[F[_]: Applicative](
     spanNamer: Http4sSpanNamer = Http4sSpanNamer.methodWithPath,
     requestFilter: Http4sRequestFilter = Http4sRequestFilter.allowAll,
     dropHeadersWhen: CaseInsensitiveString => Boolean = Headers.SensitiveHeaders.contains,
+    errorHandler: ErrorHandler = ErrorHandler.empty,
   )(k: ResourceKleisli[F, SpanParams, Span[F]]): ResourceKleisli[F, Request_, Span[F]] =
     Kleisli { req =>
       val filter = requestFilter.lift(req).getOrElse(true)
       lazy val headers = Http4sHeaders.converter.from(req.headers)
-      val spanResource = if (filter) k.run((spanNamer(req), SpanKind.Server, headers)) else Span.noop[F]
+      val spanResource =
+        if (filter) k.run((spanNamer(req), SpanKind.Server, headers, errorHandler.orElse(messageFailureHandler)))
```

Comment:
@catostrophe what do you think of the fallback precidence here? If someone creates a partial function handling all `Throwable`s it will override the message failure handler.

Reply:
@janstenpickle I think this is fine since it's overridable and under the control of a user.